server/datastore/mysql/migrations/tables/20230328091816_EnsureUniformCollation.go

```diff
@@ -3,14 +3,170 @@ package tables
 import (
 	"bytes"
 	"database/sql"
+	"encoding/json"
 	"fmt"
 	"text/template"
 
 	"github.com/jmoiron/sqlx"
 )
 
 func init() {
-	MigrationClient.AddMigration(Up_20230315104937, Down_20230315104937)
+	MigrationClient.AddMigration(Up_20230328091816, Down_20230328091816)
+}
+
+func fixupSoftware(tx *sql.Tx, collation string) error {
+	rows, err := tx.Query(`
+         SELECT
+           COUNT(*) as total,
+           CONCAT('[', GROUP_CONCAT(id SEPARATOR ','), ']') as ids
+         FROM software
+         GROUP BY ` +
+		fmt.Sprintf("`version` COLLATE %s,", collation) +
+		fmt.Sprintf("`release` COLLATE %s,", collation) +
+		fmt.Sprintf(`name      COLLATE %s,
+		source    COLLATE %s,
+		vendor    COLLATE %s,
+		arch      COLLATE %s
+		HAVING total
+		COLLATE %s`, collation, collation, collation, collation, collation))
+
+	if err != nil {
+		return err
+	}
+
+	defer rows.Close()
+	var idGroups [][]uint
+	for rows.Next() {
+		var rawIDs json.RawMessage
+		var total int
+		if err := rows.Scan(&total, &rawIDs); err != nil {
+			return err
+		}
+		var ids []uint
+		if err := json.Unmarshal(rawIDs, &ids); err != nil {
+			return err
+		}
+		idGroups = append(idGroups, ids)
+	}
+
+	for _, ids := range idGroups {
+		for i := 1; i < len(ids); i++ {
+			if _, err := tx.Exec("DELETE FROM software_cve WHERE software_id = ?", ids[i]); err != nil {
+				return err
+			}
+			if _, err := tx.Exec("DELETE FROM software_host_counts WHERE software_id = ?", ids[i]); err != nil {
+				return err
+			}
+			if _, err := tx.Exec("UPDATE host_software SET software_id = ? WHERE software_id = ?", ids[0], ids[i]); err != nil {
```

Comment:
Is it possible for this to generate duplicate `host_software` entries for a given host? If so, any idea whether that could cause problems?

Reply:
this is a good point, I don't foresee any problems, but since these are dupes, it's probably better to `DELETE` them, what do you think?

Reply:
I think there are tradeoffs to either approach but I feel slightly better about deleting.